File: 1474-longest-zigzag-path-in-a-binary-tree/longest-zigzag-path-in-a-binary-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def longestZigZag(self, root: Optional[TreeNode]) -> int:
        res = 0
        def dfs(node, dir, steps):
            nonlocal res

            if not node:
                return 0
            
            res = max(steps, res)
            if dir == "right":
                dfs(node.left, "left", steps + 1)
                dfs(node.right, "right", 1)
            if dir == "left":
                dfs(node.left, "left", 1)
                dfs(node.right, "right", steps + 1)
        
        dfs(root, "left", 0)
        return res

File: 1474-longest-zigzag-path-in-a-binary-tree/longest-zigzag-path-in-a-binary-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int longestZigZag(TreeNode* root) {
        stack<tuple<TreeNode*, int, bool>> stk;
        stk.push(make_tuple(root, 0, false));
        int res = 0;
        while(stk.size() > 0) {
            auto cur = stk.top();
            stk.pop();
            TreeNode* node = get<0>(cur);
            int steps = get<1>(cur);
            bool fromLeft = get<2>(cur);
            res = max(res, steps);
            if(node->left) {
                if(fromLeft) stk.push(make_tuple(node->left, 1, true));
                else stk.push(make_tuple(node->left, steps + 1, true));
            }
            if(node->right) {
                if (fromLeft) stk.push(make_tuple(node->right, steps + 1, false));
                else stk.push(make_tuple(node->right, 1, false));
            }
        }
        return res;
    }
};